apollo-router/src/spec/schema.rs

```diff
@@ -38,24 +42,271 @@ pub(crate) struct Schema {
+        fn parse(schema: &str, _configuration: &Configuration) -> Result<Schema, SchemaError> {
+            let mut compiler = ApolloCompiler::new();
+            compiler.create_schema(
+                include_str!("introspection_types.graphql"),
+                "introspection_types.graphql",
             );
+            let id = compiler.create_schema(schema, "schema.graphql");
 
-            parse(&api_schema, configuration)
+            let ast = compiler.db.ast(id);
+
+            // Trace log recursion limit data
+            let recursion_limit = ast.recursion_limit();
+            tracing::trace!(?recursion_limit, "recursion limit data");
+
+            // TODO: run full compiler-based validation instead?
+            let errors = ast.errors().cloned().collect::<Vec<_>>();
+            if !errors.is_empty() {
+                let errors = ParseErrors {
+                    raw_schema: schema.to_string(),
+                    errors,
+                };
+                errors.print();
+                return Err(SchemaError::Parse(errors));
+            }
+
+            fn as_string(value: &hir::Value) -> Option<&String> {
+                if let hir::Value::String(string) = value {
+                    Some(string)
+                } else {
+                    None
+                }
+            }
+
+            let mut subgraphs = HashMap::new();
+            // TODO: error if not found?
+            if let Some(join_enum) = compiler.db.find_enum_by_name("join__Graph".into()) {
+                for (name, url) in join_enum
+                    .enum_values_definition()
+                    .iter()
+                    .filter_map(|value| {
+                        let join_directive = value
+                            .directives()
+                            .iter()
+                            .find(|directive| directive.name() == "join__graph")?;
+                        let name = as_string(join_directive.argument_by_name("name")?)?;
+                        let url = as_string(join_directive.argument_by_name("url")?)?;
+                        Some((name, url))
+                    })
+                {
+                    if url.is_empty() {
+                        return Err(SchemaError::MissingSubgraphUrl(name.clone()));
+                    }
+                    let url = Uri::from_str(url)
+                        .map_err(|err| SchemaError::UrlParse(name.clone(), err))?;
+                    if subgraphs.insert(name.clone(), url).is_some() {
+                        return Err(SchemaError::Api(format!(
+                            "must not have several subgraphs with same name '{}'",
+                            name
+                        )));
+                    }
+                }
+            }
+
+            let object_types: HashMap<_, _> = compiler
+                .db
+                .object_types()
+                .iter()
+                .map(|(name, def)| (name.clone(), (&**def).into()))
```

Comment:
Is there a simpler way to write this?

Reply:
I don’t think so, unless the conversion is changed from `impl From<&hir::ObjectTypeDefinition> for ObjectType` to `impl From<&Arc<hir::ObjectTypeDefinition>> for ObjectType` which I’m not a fan of.